cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/SystemAppTask.java

```diff
@@ -16,54 +16,120 @@
     try (CloseableClassLoader artifactClassLoader =
            artifactRepository.createArtifactClassLoader(artifactDetail.getDescriptor().getLocation(),
                                                         classLoaderImpersonator)) {
+      RunnableTaskRequest taskRequest = GSON.fromJson(context.getParam(), RunnableTaskRequest.class);
       String taskClassName = taskRequest.getClassName();
       Class<?> clazz = artifactClassLoader.loadClass(taskClassName);
       if (!(RunnableTask.class.isAssignableFrom(clazz))) {
         throw new ClassCastException(String.format("%s is not a RunnableTask", taskClassName));
       }
-
+      TaskSystemAppContext taskSystemAppContext = buildTaskSystemAppContext(injector, systemAppNamespace,
+                                                                            systemAppArtifactId, artifactClassLoader);
       RunnableTask runnableTask = (RunnableTask) injector.getInstance(clazz);
       RunnableTaskContext runnableTaskContext = RunnableTaskContext.getBuilder().
         withParam(taskRequest.getParam()).
-        withArtifactClassLoader(artifactClassLoader).build();
+        withTaskSystemAppContext(taskSystemAppContext).build();
       runnableTask.run(runnableTaskContext);
       context.writeResult(runnableTaskContext.getResult());
+      taskSystemAppContext.releaseResources();
+    }
+  }
+
+  private TaskSystemAppContext buildTaskSystemAppContext(Injector injector, String systemAppNamespace,
+                                                         ArtifactId artifactId, ClassLoader artifactClassLoader) {
+    DatasetFramework dsFramework = injector.getInstance(DatasetFramework.class);
+    SecureStoreManager secureStoreManager = injector.getInstance(SecureStoreManager.class);
+    MessagingService messagingService = injector.getInstance(MessagingService.class);
+    NamespaceQueryAdmin namespaceQueryAdmin = injector.getInstance(NamespaceQueryAdmin.class);
+    PreferencesFetcher preferencesFetcher = injector.getInstance(PreferencesFetcher.class);
+    PluginFinder pluginFinder = injector.getInstance(PluginFinder.class);
+    RetryStrategy retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + ".");
+    DiscoveryServiceClient discoveryServiceClient = injector.getInstance(DiscoveryServiceClient.class);
+    SecureStore secureStore = injector.getInstance(SecureStore.class);
+    return
+      new DefaultTaskSystemAppContext(cConf, dsFramework, secureStoreManager, messagingService,
+                                      retryStrategy, namespaceQueryAdmin, preferencesFetcher, pluginFinder,
+                                      discoveryServiceClient, secureStore, systemAppNamespace, artifactId,
+                                      artifactClassLoader);
+  }
+
+  private void streamArtifactIfRequired(Id.Artifact artifactId, Location artifactLocation,
```

Comment:
Is the location local?? Is suppose this logic would change with the sidecar caching mechanism?

Reply:
Yes, this will use the sidecar container end point to fetch the location.